engine/verification/requester/requester.go

```diff
@@ -0,0 +1,248 @@
+func (e *Engine) handleChunkDataPack(originID flow.Identifier, chunkDataPack *flow.ChunkDataPack, collection *flow.Collection) {
+	chunkID := chunkDataPack.ChunkID
+	collectionID := collection.ID()
+	lg := e.log.With().
+		Hex("chunk_id", logging.ID(chunkID)).
+		Hex("collection_id", logging.ID(collectionID)).
+		Logger()
+	lg.Debug().Msg("chunk data pack received")
+
+	// makes sure we still need this chunk, and we will not process duplicate chunk data packs.
+	removed := e.pendingRequests.Rem(chunkID)
+	if !removed {
+		lg.Debug().Msg("chunk request status not found in mempool to be removed, dropping chunk")
+		return
+	}
+
+	e.handler.HandleChunkDataPack(originID, chunkDataPack, collection)
+	lg.Info().Msg("successfully sent the chunk data pack to the handler")
+}
+
+// Request receives a chunk data pack request and adds it into the pending requests mempool.
+func (e *Engine) Request(request *verification.ChunkDataPackRequest, targets flow.IdentityList) {
+	status := &verification.ChunkRequestStatus{
+		ChunkDataPackRequest: request,
+		Targets:              targets,
+	}
+	added := e.pendingRequests.Add(status)
+	e.log.Info().
+		Hex("chunk_id", logging.ID(request.ChunkID)).
+		Uint64("block_height", request.Height).
+		Int("agree_executors", len(request.Agrees)).
+		Int("disagree_executors", len(request.Disagrees)).
+		Bool("added_to_pending_requests", added).
+		Msg("chunk data pack request arrived")
+}
+
+// onTimer should run periodically, it goes through all pending requests, and requests their chunk data pack.
+// It also retries the chunk data request if the data hasn't been received for a while.
+func (e *Engine) onTimer() {
+	pendingReqs := e.pendingRequests.All()
+
+	e.log.Debug().
+		Int("total", len(pendingReqs)).
+		Msg("start processing all pending chunk data requests")
+
+	for _, request := range pendingReqs {
+		lg := e.log.With().
+			Hex("chunk_id", logging.ID(request.ID())).
+			Uint64("block_height", request.Height).
+			Logger()
+
+		// if block has been sealed, then we can finish
+		sealed, err := e.blockIsSealed(request.Height)
+		if err != nil {
+			lg.Fatal().Err(err).Msg("could not determine whether block has been sealed")
+			continue
+		}
+		if sealed {
+			removed := e.pendingRequests.Rem(request.ID())
+			e.handler.NotifyChunkDataPackSealed(request.ID())
+			lg.Info().
+				Bool("removed", removed).
+				Msg("drops requesting chunk of a sealed block")
+			continue
+		}
+
+		err = e.requestChunkDataPack(request)
+		if err != nil {
+			lg.Warn().Err(err).Msg("could not request chunk data pack")
+			continue
+		}
+
+		updated := e.pendingRequests.IncrementAttempt(request.ID())
+		if !updated {
+			lg.Debug().Msg("could not update dispatched request, it has been resolved")
+		}
+
+		lg.Info().Msg("chunk data pack requested")
+	}
+}
+
+// blockIsSealed returns true if the block at specified height is sealed.
+func (e Engine) blockIsSealed(height uint64) (bool, error) {
+	lastSealed, err := e.state.Sealed().Head()
```

Comment:
fixed in https://github.com/onflow/flow-go/pull/633/commits/d0af0d82860b03d70881e4e17d70e2978ca6e2e1, just a heads up that it extends the false-positive window of block of a chunk is getting sealed once its turn is reached in the loop. However, since we are evaluating that using the local cached variable, it falsely shows chunk unsealed, and requests it (of course getting caught on requester timer loop).